src/Engine/FloodFillFunctor.java

```diff
@@ -0,0 +1,69 @@
+package Engine;
+
+import Terrain.GameMap;
+import Units.Unit;
+import Units.MoveTypes.MoveType;
+
+/**
+ * Provides a stateless interface for determining traversability
+ * Used in {@link #findShortestPath(XYCoord, Unit, int, int, GameMap, boolean) findShortestPath()} and {@link #findPossibleDestinations(XYCoord, Unit, GameMap, boolean) findPossibleDestinations()}
+ */
+public interface FloodFillFunctor
+{
+  /**
+   * Determines whether the Location (x, y), can be traveled through.
+   * @param ignoreUnits If set, this function will ignore enemy-unit presence.
+   */
+  int getRemainingFillPower(GameMap map, int initialFillPower, XYCoord from, XYCoord to);
+  /**
+   * Determines whether the Location (x, y), is a valid destination.
+   */
+  boolean canEnd(GameMap map, XYCoord end);
+
+
+  /** Acts as the default way of calculating unit mobility */
+  public static class BasicMoveFillFunctor implements FloodFillFunctor
+  {
+    public final Unit unit;
+    public final boolean includeOccupiedDestinations;
+    public final boolean canTravelThroughEnemies;
+
+    public BasicMoveFillFunctor(Unit mover, boolean includeOccupied, boolean canTravelThroughEnemies)
+    {
+      unit = mover;
+      this.includeOccupiedDestinations = includeOccupied;
+      this.canTravelThroughEnemies = canTravelThroughEnemies;
+    }
+
+    public int getRemainingFillPower(GameMap map, int initialFillPower, XYCoord from, XYCoord to)
+    {
+      int cost = findMoveCost(from, to, map);
```

Comment:
Shouldn't we check that the map location is valid before we try to find the move cost?
It seems like we'd want
```
if( !map.locationValid(to) )
{...}
else if( !canTravel ... )
{...}
else
  cost = findMoveCost()
```